text/0000-native-class-constructor-update.md

```diff
@@ -0,0 +1,220 @@
+subclasses have been assigned. Revisiting our previous example, the following
+two class declarations would effectively be equivalent:
+
+```js
+const Foo = EmberObject.extend({
+  bar: 'baz'
+});
+
+class Foo extends EmberObject {
+  bar = 'baz';
+}
+```
+
+Much easier to codemod! There are other subtle differences between native class
+fields and EmberObject properties, such as the fact that class fields are
+assigned each time a class is initialized, but these are easier to work around.
+
+### Injections and the `init` hook
+
+One side effect of this change is that injections will not be available on the
+class instance during the `constructor` phase. The ideal behavior here is
+outside of the scope of this RFC, and is something that should be discussed in
+future RFCs. For the time being, users can still rely on the `init` hook, which
+will continue to be called after all injections and properties have been
+assigned to the instance.
+
+### `new EmberObject()`
+
+It was previously possible to use `new` syntax with EmberObject. While this
+was not considered public API, it has technically worked and been under test
+since the early days of Ember, and may fall under the category of intimate API.
+
+We can continue to support this behavior in a backwards compatible way while
+deprecating it with one final tweak to the change above:
+
+```js
+const DEFER_INIT = new Symbol();
+
+function initialize(instance, props) {
+  Object.assign(instance, props);
+  instance.init();
+}
+
+class EmberObject {
+  constructor(props, maybeDefer) {
+    // ..class setup things
+
+    if (maybeDefer === DEFER_INIT) {
+       return this;
+    }
+
+    deprecate('using new with EmberObject has been deprecated', true);
+    initialize(this, props);
+  }
+
+  static create(props) {
+    let instance = new this(props, DEFER_INIT);
+    initialize(instance, props);
+
+    return instance;
+  }
+}
+```
+
+## How we teach this
+
+If this PR is accepted, most of the major issues with classes will have been
+resolved. We can begin working on a codemod to make converting easier, and move
+toward officially making native classes a finalized part of the public API of
+Ember. Pending decorators and class fields moving to a late enough stage in the
+TC39 process, we can also begin converting the guides to use native class
+syntax.
+
+We can document the exact behavior of the new constructor in the API docs for
+EmberObject. Most details won't have to change since this change only affects
+native class syntax, which has not been documented much officially. We can also
+demonstrate the behaviors of classes throughout the guides and API docs.
+
+One thing we should make clear is that EmberObject will likely be deprecated
+in the near future, and that ideally for non-Ember classes (things that aren't
+Components, Services, etc.) users should drop EmberObject altogether and use
+native classes only. For the purposes of codemodding, however, this will not be
```

Comment:
This last sentence is a bit unclear; I actually have no idea what it means. Could you reword?

Reply:
Coming back to it, I think it was pretty unclear too so I'm going to drop it. Basically, it's not possible to easy codemod a class which was base on `EmberObject` to one which is _not_ based on `EmberObject`, for a number of reasons:

- Common usage of static methods (`create` is used everywhere, and tracking down and converting each usage would be _very_ difficult without types in non TS apps)
- Mixins. If they're used once in the inheritance tree, we can't convert.

The point is, we shouldn't be recommending that people use `EmberObject` at all now that we have native classes, and we should guide users toward converting any existing classes.